### PR TITLE
[Clang][Sema] Use the correct injected template arguments for partial specializations when collecting multi-level template argument lists

### LINKED ISSUE #112222: [clang++] Crash when parsing out-of-class definition of member function with trailing requires clause
I'm observing crashes when compiling the following snipped with clang trunk. Clang 19 doesn't crash.

``` c++
#include <cstdint>

template<typename T>
class TExprBase;

template<std::size_t dim, std::size_t rank, template<std::size_t, std::size_t> typename T>
class TExprBase<T<dim, rank>>
{
public:
  void Curl() const requires (dim == 3u && rank == 1u);
};

template<std::size_t dim, std::size_t rank, template<std::size_t, std::size_t> typename T>
void TExprBase<T<dim, rank>>::Curl() const requires (dim == 3u && rank == 1u)
{
}
```

### PR DESCRIPTION
**Pull request by sdkrystian**
After #111852 refactored multi-level template argument list collection, the following results in a crash:
```cpp
template<typename T, bool B>
struct A;

template<bool B>
struct A<int, B>
{
    void f() requires B;
};

template<bool B>
void A<int, B>::f() requires B { } // crash here
```

This happens because when collecting template arguments for constraint normalization from a partial specialization, we incorrectly use the template argument list of the partial specialization. We should be using the template argument list of the _template-head_ (as defined in [[temp.arg.general] p2](http://eel.is/c++draft/temp.arg.general#2)). Fixes #112222.

Still need to add tests.

### CODE REVIEW

**Review by erichkeane — APPROVED** [review]
Change itself seems reasonable, so approve, pending a good couple of tests.

**Review by mizvekov — APPROVED** [review]

File: clang/include/clang/AST/DeclTemplate.h [hunk 1]
```diff
@@ -2085,7 +2085,9 @@ class ClassTemplateSpecializationDecl : public CXXRecordDecl,
 class ClassTemplatePartialSpecializationDecl
   : public ClassTemplateSpecializationDecl {
   /// The list of template parameters
-  TemplateParameterList* TemplateParams = nullptr;
+  TemplateParameterList *TemplateParams = nullptr;
+
+  TemplateArgument *InjectedArgs = nullptr;
```
Comment (zyn0217) [inline clang/include/clang/AST/DeclTemplate.h:2090]: Do we need serialization?

**Review by zyn0217 — APPROVED** [review]
LG modulo one question. Thanks for the prompt fix.

File: clang/include/clang/AST/DeclTemplate.h (hunk 1 above)
Comment (mizvekov) [reply clang/include/clang/AST/DeclTemplate.h:2090]: This is just for catching the injected arguments. If we don't serialize these, they will be just rebuilt anyway.

Comment (sdkrystian) [reply clang/include/clang/AST/DeclTemplate.h:2092]: Yeah, like @mizvekov said we don't serialize injected template arguments

**Pull request merged**